### PR TITLE
S3 datastore support

### PR DESCRIPTION
**Pull request by tv42**
For your reviewing pleasure, do not merge yet. Sorry for the noise in the pull request, this is on top of the `pin` branch, start reading from the most recent commit with "pin:" in the commit message. Once pinning is fully merged, I'll rebase this and ask for a merge, but for now living on top of that branch causes less noise in the datastore.

Note that S3 objects names are hex, because datastore keys can be binary. So that means S3 objects like `2f626c6f636b732f122014d8c7e2382fe98c052b96b7a86daa9b176fde6324cbfdd37314617dc1058a7d`.

### CODE REVIEW

**Comment by whyrusleeping** [comment]
@jbenet dont you dare git push each this one, i want to run other tests this month.

**Comment by whyrusleeping** [comment]
@tv42 do you have any thoughts on query for the s3datastore? thats gonna be the hard part i think.

Also, how hard would it be to run a local 's3' server for testing this code?

Most of this looks good so far.

**Comment by tv42** [comment]
@whyrusleeping I don't expect S3 to bring any special difficulties to Query. I do want to whack the Query API with a mallet; whether that happens as part of the continuation of this or not depends on whether I come up with a good enough replacement API soon enough. If not, s3 Query will look a lot like flatfs Query, except it could support more (unused) features.

I'm not aware of any "tiny s3 clones" worth recommending, and didn't want to start writing one on a whim. Making a true S3 clone is ridiculously hard (been there; https://github.com/ceph/s3-tests/ ), though this code uses a small enough of a fraction that httptest.Server might be plenty enough.

Also, remember that it's currently only 70 lines of code and 83 lines to configure that (most of that validating the JSON, for better error reporting), and maybe 60-100 LoC for a simple Query implementation. And it's pretty much all very simple code.

Perhaps some of the sharness stuff could be run in a special mode, where they talk to actual AWS S3?

**Comment by whyrusleeping** [comment]
> Perhaps some of the sharness stuff could be run in a special mode, where they talk to actual AWS S3?

Maybe... that's probably okay to do if we make sure they only run when a given env var (or set of env vars) is set

File: repo/config/datastore.go [hunk 1]
```diff
@@ -1,16 +1,83 @@
...
+	datastore.ThreadSafeDatastore
+	Close() error
+}
+
+// DSOpener is  an interface that lets the caller open a datastore.
+//
+// All values of DSOpener must have a JSON representation that is
+// semantically equivalent to the Datastore section of the config
+// file that they were instantiated by, including the "Type" key.
+type DSOpener interface {
```
Comment (jbenet) [inline repo/config/datastore.go]: good idea :+1:

**Comment by jbenet** [comment]
@tv42 this PR is great. +1 everything LGTM. sorry for the huge delay on my end.

> I do want to whack the Query API with a mallet; whether that happens as part of the continuation of this or not depends on whether I come up with a good enough replacement API soon enough. If not, s3 Query will look a lot like flatfs Query, except it could support more (unused) features.

Either sounds good to me.

> though this code uses a small enough of a fraction of features that httptest.Server might be plenty enough.

Yeah i think this works for normal tests. And, there's many S3 mock things out there that already implement the exact API for us (not that it's complicated...)

> > Perhaps some of the sharness stuff could be run in a special mode, where they talk to actual AWS S3?
> Maybe... that's probably okay to do if we make sure they only run when a given env var (or set of env vars) is set

Yeah, we can have a special set of sharness tests that can be run against S3. I wouldn't want to put this on CI for all PRs. (could make a simple side-repo with the right config that pulls go-ipfs and tests on S3, and make sure to pull to this repo before merging to stable. would need a way to make credentials private or whatever)

---

@whyrusleeping 
> @jbenet dont you dare git push each this one, i want to run other tests this month.

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: ef347684bdb626f2234323336d51111e8858333a
  * Invalid signoff. Commit message must end with **License: MIT
Signed-off-by: .* <.*>**

[Guidelines](https://github.com/ipfs/community/blob/master/contribution-guidelines.md#commit-messages) are available to help. Your feedback on GitCop is welcome on [this issue](https://github.com/ipfs/infrastructure/issues/23)

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: ef347684bdb626f2234323336d51111e8858333a
  * Invalid signoff. Commit message must end with **License: MIT
Signed-off-by: .* <.*>**

[Guidelines](https://github.com/ipfs/community/blob/master/contribution-guidelines.md#commit-messages) are available to help. Your feedback on GitCop is welcome on [this issue](https://github.com/ipfs/infrastructure/issues/23)

---
This message was auto-generated by https://gitcop.com

**Comment by tv42** [comment]
Gitcop is wrong (https://github.com/ipfs/infrastructure/issues/46), and my commits have been amended to have the right trailer.

Unit tests pass, sharness passes.

This is on top of dev0.4.0 (with a fix for a random bug I found in dev0.4.0) and ready to roll! Paging @jbenet @whyrusleeping

**Comment by tv42** [comment]
I have no idea about those circleci & travis errors in https://circleci.com/gh/ipfs/go-ipfs/640 https://travis-ci.org/ipfs/go-ipfs/jobs/71014668 but none of that happens locally and none of it seems to be related to these changes.

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 82afdb112c85fb03b607a5570dfaf1de8b1e7f09
  * Invalid signoff. Commit message must end with **License: MIT
Signed-off-by: .* <.*>**

[Guidelines](https://github.com/ipfs/community/blob/master/contribution-guidelines.md#commit-messages) are available to help. Your feedback on GitCop is welcome on [this issue](https://github.com/ipfs/infrastructure/issues/23)

---
This message was auto-generated by https://gitcop.com

**Comment by tv42** [comment]
Rebased again to get on top of latest dev0.4.0.

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 82afdb112c85fb03b607a5570dfaf1de8b1e7f09
  * Invalid signoff. Commit message must end with **License: MIT
Signed-off-by: .* <.*>**

[Guidelines](https://github.com/ipfs/community/blob/master/contribution-guidelines.md#commit-messages) are available to help. Your feedback on GitCop is welcome on [this issue](https://github.com/ipfs/infrastructure/issues/23)

---
This message was auto-generated by https://gitcop.com

**Comment by tv42** [comment]
Closing in favor of https://github.com/ipfs/go-ipfs/pull/1488 because that one targets dev0.4.0. This one is from before the dev branch existed.

**Pull request closed without merging**